Sources/Networking/DecodableAlamofire/DataRequest+Decodable.swift

```diff
@@ -52,7 +52,7 @@ final class DataKeyPathSerializer<SerializedObject: Decodable>: DataResponseSeri
             throw AFError.responseSerializationFailed(reason: .decodingFailed(error: AlamofireDecodableError.emptyKeyPath))
         }
         
-        let json = try JSONResponseSerializer().serialize(request: nil, response: response, data: data, error: nil)
+        let json = try DecodableResponseSerializer<[String: SerializedObject]>().serialize(request: nil, response: response, data: data, error: nil)
```

Comment:
I've found an issue where, unfortunately, this will not work.

If you have this JSON:
```json
{
    "id": "1234",
    "greeting": "Hello",
    "location": null
}
```
and on the call site you define it like this:
```swift
apiService
    .request(
        String.self,
        decodingKeyPath: "id"
    )
```

we will get an error that we `defined type String, but found nil in "location"`.

The way how we can handle this is by changing a few things:
1. Change `DataResponseSerializerProtocol` to `ResponseSerializer` (to access `emptyResponseAllowed`)
2. Replace `DecodableResponseSerializer` with `JSONSerialization`
3. Add check if `Data` is empty.

```swift
final class DataKeyPathSerializer<SerializedObject: Decodable>: ResponseSerializer {

    private let keyPath...
    ...
   
    init...
    
    func serialize(request: URLRequest?, response: HTTPURLResponse?, data: Data?, error: Error?) throws -> SerializedObject {
        // If there is an error, throw immediately
        if let error = error {
            throw error
        }

        // In case there is no keypath fallback to default Alamofire parsing
        guard let keyPath = keyPath else {
            let data = try DataResponseSerializer().serialize(request: nil, response: response, data: data, error: nil)
            let object = try self.decoder.decode(SerializedObject.self, from: data)
            return object
        }

        if keyPath.isEmpty {
            throw AFError.responseSerializationFailed(reason: .decodingFailed(error: AlamofireDecodableError.emptyKeyPath))
        }

        /* UPDATED PART */
        guard let data = data, !data.isEmpty else {
            guard emptyResponseAllowed(forRequest: request, response: response) else {
                throw AFError.responseSerializationFailed(reason: .inputDataNilOrZeroLength)
            }
            throw APIError.emptyResponseBody // <-- Define this
        }

        let json = try JSONSerialization.jsonObject(with: data, options: .allowFragments)
        ...
```

Inside Alamofire, the `Define this` part is returning `NSNull()`, which we are then checking inside this file, and in general, we are entering the `else` block where we are throwing `AFError.responseSerializationFailed(reason: .decodingFailed(error: AlamofireDecodableError.invalidKeyPath))` error.

Should we do the same in this `Define this` line? (easiest approach)
Should we define the return type of the `serialize` method as optional (i.e., `SerializedObject?`) and then return `nil`? (maybe the more correct way, as we don't have the Data, but it would break everyone's in-app logic throughout the app)?

cc @Truba @hhrvoic

Reply:
Shouldn't the app be responsible to define what is mandatory/optional? If app asks for a String in example above, it basically tells to the service that it has to exist. If it can be optional, app can ask for an optional? I believe in the example you meant

```
apiService
    .request(
        String.self,
        decodingKeyPath: "location"
    )
```
?

Reply:
The example was correct, the app requests `"id"`, so I don't want to decode everything, I just need one value from the entire response. 

The same will be if we have this response:
```json
{
    "user": {
        "name": "John",
        "age": 30
    },
    "id": 422,
    "unique": true
}
```
and we don't want to decode ID and Unique values, so we define it like this:
```swift
apiService
    .request(
        UserModel.self,
        decodingKeyPath: "user"
)
```

If we keep using `DecodableResponseSerializer<[String: SerializedObject]>` (where SerializedObject is `UserModel`), we will get an error: "Defined UserModel, but found Int in id" as that serializer is trying to decode entire JSON as `[String: UserModel]`.

For this keypath decoding to work, we must decode/encode it twice (a known issue from before, not relevant for this PR). As we have defined that `DecodableResponseSerializer` is JSON/dictionary of `[String : UserModel]`, our logic will fail as `"id"` and `"unique"` types are not `UserModel`.

My proposal was that we should use `JSONSerialization.jsonObject` instead as this will give us `Any` that we can then cast to `(json as AnyObject).value(forKeyPath: keyPath)` (where keyPath is a value passed in `.request`), and then we can check if that nested value is `UserModel`.

As `JSONSerialization.jsonObject` requires non-optional/not empty `data`, we must check if the data is there. If it isn't, we have to throw _something_. We can throw our error (e.g., `APIError.emptyResponseBody`), or we can use the error defined inside Alamofire (as we already do in that file, e.g., `AFError.responseSerializationFailed(reason: .inputDataNilOrZeroLength)`)

So my question is if using `JSONSerialization.jsonObject` is an ok approach? 
What should we throw in case data is not there?

Reply:
Solution seems ok 👍 . If this is still a `ResponseSerializer` (Alamofire's type), I have no objections using its existing type for errors which already has all the cases covered (`AFError`).

Reply:
After internal discussion, we decided that `DataKeyPathSerializer` will conform to `ResponseSerializer` and that we will add this part of the code:
```swift
        guard let data = data, !data.isEmpty else {
             guard emptyResponseAllowed(forRequest: request, response: response) else {
                 throw AFError.responseSerializationFailed(reason: .inputDataNilOrZeroLength)
             }
             throw AFError.responseSerializationFailed(reason: .decodingFailed(error: AlamofireDecodableError.invalidKeyPath))
         }

         let json = try JSONSerialization.jsonObject(with: data, options: .allowFragments)
```

@ssvetina can you please make those changes in this PR? Thank you :)

Reply:
@jabou Ofcourse, I'll add the suggested approach. Thanks guys :)